ecc/bls12-377/fr/iop/utils.go

```diff
@@ -0,0 +1,378 @@
+	_p.Info.Status = Unlocked
+	d.FFTInverse(_p.Coefficients, fft.DIF)
+	return &_p
+}
+
+// LAGRANGE REGULAR UNLOCKED
+func toCanonical5(p *Polynomial, d *fft.Domain) *Polynomial {
+	d.FFTInverse(p.Coefficients, fft.DIF)
+	p.Info.Basis = Canonical
+	p.Info.Layout = BitReverse
+	return p
+}
+
+// LAGRANGE BITREVERSE LOCKED
+func toCanonical6(p *Polynomial, d *fft.Domain) *Polynomial {
+	_p := copyPoly(*p)
+	_p.Info.Basis = Canonical
+	_p.Info.Layout = Regular
+	_p.Info.Status = Unlocked
+	d.FFTInverse(_p.Coefficients, fft.DIT)
+	return &_p
+}
+
+// LAGRANGE BITREVERSE UNLOCKED
+func toCanonical7(p *Polynomial, d *fft.Domain) *Polynomial {
+	d.FFTInverse(p.Coefficients, fft.DIT)
+	p.Info.Basis = Canonical
+	p.Info.Layout = Regular
+	return p
+}
+
+// LAGRANGE_COSET REGULAR LOCKED
+func toCanonical8(p *Polynomial, d *fft.Domain) *Polynomial {
+	_p := copyPoly(*p)
+	_p.Info.Basis = Canonical
+	_p.Info.Layout = BitReverse
+	_p.Info.Status = Unlocked
+	d.FFTInverse(_p.Coefficients, fft.DIF, true)
+	return &_p
+}
+
+// LAGRANGE_COSET REGULAR UNLOCKED
+func toCanonical9(p *Polynomial, d *fft.Domain) *Polynomial {
+	p.Info.Basis = Canonical
+	p.Info.Layout = BitReverse
+	d.FFTInverse(p.Coefficients, fft.DIF, true)
+	return p
+}
+
+// LAGRANGE_COSET BITREVERSE LOCKED
+func toCanonical10(p *Polynomial, d *fft.Domain) *Polynomial {
+	_p := copyPoly(*p)
+	_p.Info.Basis = Canonical
+	_p.Info.Layout = Regular
+	d.FFTInverse(_p.Coefficients, fft.DIT, true)
+	return &_p
+}
+
+// LAGRANGE_COSET BITREVERSE UNLOCKED
+func toCanonical11(p *Polynomial, d *fft.Domain) *Polynomial {
+	p.Info.Basis = Canonical
+	p.Info.Layout = Regular
+	d.FFTInverse(p.Coefficients, fft.DIT, true)
+	return p
+}
+
+var _toCanonical [12]modifier = [12]modifier{
+	toCanonical0,
+	toCanonical1,
+	toCanonical2,
+	toCanonical3,
+	toCanonical4,
+	toCanonical5,
+	toCanonical6,
+	toCanonical7,
+	toCanonical8,
+	toCanonical9,
+	toCanonical10,
+	toCanonical11,
+}
+
+// toCanonical changes or returns a copy of p (according to its
+// status, Locked or Unlocked), or modifies p to put it in Lagrange
+// basis. The result is not bit reversed.
+func toCanonical(p *Polynomial, d *fft.Domain) *Polynomial {
+	return _toCanonical[getShapeID(*p)](p, d)
+}
+
+//----------------------------------------------------
+// exp functions until 5
+
+func exp0(x fr.Element) fr.Element {
+	var res fr.Element
+	res.SetOne()
+	return res
+}
+
+func exp1(x fr.Element) fr.Element {
+	return x
+}
+
+func exp2(x fr.Element) fr.Element {
+	return *x.Square(&x)
+}
+
+func exp3(x fr.Element) fr.Element {
+	var res fr.Element
+	res.Square(&x).Mul(&res, &x)
+	return res
+}
+
+func exp4(x fr.Element) fr.Element {
+	x.Square(&x).Square(&x)
+	return x
+}
+
+func exp5(x fr.Element) fr.Element {
+	var res fr.Element
+	res.Square(&x).Square(&res).Mul(&res, &x)
+	return res
+}
+
+// doesn't return any errors, it is a private method, that
+// is assumed to be called with correct arguments.
+func smallExp(x fr.Element, n int) fr.Element {
```

Comment:
unused? 
if used, I think it's less verbose to init 5 big int to avoid many allocations, do a switch on n and call fr.Element.Exp(.., bigINt), perf should be OK.